specification/Metadata/README.md

```diff
@@ -0,0 +1,1130 @@
+
+Array types are straightforward to encode into JSON: instead of a single array of values, use an array of arrays. When the array has a fixed size (i.e. `componentCount` is defined), each of the inner arrays must have the same length. Otherwise, the arrays are assumed to be variable-length.
+
+The following example shows the usage for both fixed and variable size arrays:
+
+```json
+{
+  "classes": {
+    "basicTypes": {
+      "properties": {
+        "fixedArrayProperty": {
+          "type": "ARRAY",
+          "componentType": "UINT8",
+          "componentCount": 2
+        },
+        "variableSizeArrayProperty": {
+          "type": "ARRAY",
+          "componentType": "STRING",
+        }
+      }
+    }
+  },
+  "instanceTables": {
+    "basicTypesTable": {
+      "class": "basicTypes",
+      "count": 3,
+      "properties": {
+        "fixedArrayProperty": {
+          "values": [
+            [3, 45],
+            [128, 110],
+            [63, 4]
+          ]
+        },
+        "variableSizeArrayProperty": {
+          "values": [
+            ["plaster", "wood", "brick"],
+            ["steel", "glass"],
+            ["concrete"]
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+#### Blob Base64 Encoding
+
+Unlike the other data types, binary blobs do not have a straightforward JSON encoding. This is because JSON uses UTF-8 encoding, not arbitrary binary data. Raw binary values are often invalid UTF-8 sequences. This spec requires that binary blobs are Base64 encoded.
+
+Base64 is easy to transmit as it only uses ASCII characters, but is not very efficient for storage. If large blobs are needed, it is highly recommended to use the binary metadata encoding instead.
+
+```json
+{
+  "classes": {
+    "basicTypes": {
+      "properties": {
+        "blobProperty": {
+          "type": "BLOB"
+        }
+      }
+    }
+  },
+  "instanceTables": {
+    "basicTypesTable": {
+      "class": "basicTypes",
+      "count": 3,
+      "properties": {
+        "blobProperty": {
+          "values": [
+            "TG9va2luZyBmb3Igc2VjcmV0cw==",
+            "WW91IGRlY29kZWQgdGhlIG1lc3NhZ2U=",
+            "QnV0IG5vbmUgY291bGQgYmUgZm91bmQ="
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+#### Fixed-length Strings and Blobs
+
+In some cases, strings have a known, fixed length. For example, an application may require dates to be entered in `YYYYMMDD` format, which will fit in exactly 8 bytes. For this,
+the specification provides a `stringByteLength` property. In the JSON encoding, this byte length refers to the length of the string when UTF-8 encoded, regardless of how it is deplayed.
+
+`blobByteLength` is similar, but used for binary data. The length is measured as the _decoded_ binary length, not the length of the Base64 string.
+
+```json
+{
+  "classes": {
+    "basicTypes": {
+      "properties": {
+        "date": {
+          "name": "Date (YYYYMMDD)",
+          "type": "STRING",
+          "stringByteLength": 4
```

Comment:
:man_facepalming:  wait I probably have one of my diagrams wrong as well, for some reason I had "YYYYMMDD is 6 characters" in my head when really it's 8.

Fortunately since these packet diagrams are generated using `packetdiag` should be easy to fix.